File: version_3/tools/diagnose_violation_rate.py
"""Diagnostic for the persistent ``weighted_violation_rate ~= 0.96``
the prior agent observed across all soft / soft_fast Stage-2 / Stage-3
runs. Read-only tool — does not run the controller, does not need GPU.

# What the diagnostic does

Given a completed bench run dir with ``trajectory.npz``, this script
computes — under the truth params (``TRUTH_PARAMS_V5``) — three
breakdowns of where the chance-constraint indicator
``(A_traj < A_sep_per_bin)`` is firing:

1. **Per-bin evaluator** (matches what the bench reports). For each of
   the ~1344 bins (T=14d × 96 bins/day) computes A_sep at the
   instantaneous Phi and the indicator. Reports the fraction of bins
   in each regime (mono-stable healthy / collapsed / bistable).

2. **Per-day evaluator** (alternative formulation). Averages Phi over
   each day (96 bins) and computes A_sep at the daily-mean Phi; same
   for A. Reports per-day regime classification and violation rate.

3. **Active-bin only** (a third option). Restricts the indicator to
   bins where ``Phi_B + Phi_S > threshold`` (default 0.05) — i.e.
   ignores rest periods.

# Why this matters

The bursty Phi-burst expansion (`_phi_burst.py`) means most per-bin
Phi values are 0.0 for both channels. At ``Phi=(0,0)``, the v5 closed
island is mono-stable collapsed: ``A_sep = +inf``, so the indicator
fires for every rest bin regardless of what A actually is. This
mechanically produces ~96% violation when the per-bin formulation is
used on a burst schedule. Run-16 (overtrained, full HMC) was confirmed
to land at exactly 0.9643 = 1296/1344 by this mechanism.

The per-day reformulation gives a much more informative reading on the
same data — for run-16 it drops to 0.214 (3/14 days collapsed), which
matches the visual evidence that the controller is trying (and mostly
succeeding) to keep the daily-mean schedule near the (0.30, 0.30)
healthy island.

This script does NOT change the production cost function — that's a
formulation decision for Ajay to make. It only surfaces the alternative
readings on existing data.

# Usage

    cd version_3
    PYTHONPATH=.:.. python tools/diagnose_violation_rate.py \\
        outputs/fsa_v5/experiments/old_experiments/run16_stage2_soft_fast_overtrained_T14_full_hmc

If no path is given, runs against the last numbered run dir under
``outputs/fsa_v5/experiments/``.
"""

from __future__ import annotations

import os
os.environ.setdefault('JAX_ENABLE_X64', 'True')

import json
import sys
from pathlib import Path

import jax
jax.config.update('jax_enable_x64', True)
import jax.numpy as jnp
import numpy as np

from version_3.models.fsa_v5._dynamics import TRUTH_PARAMS_V5
from version_3.models.fsa_v5.control_v5 import _jax_find_A_sep

BINS_PER_DAY = 96


def _classify(asep_array: np.ndarray) -> dict:
    return {
        'healthy_mono':     int(np.sum(np.isneginf(asep_array))),
        'bistable_finite':  int(np.sum(np.isfinite(asep_array))),
        'collapsed_mono':   int(np.sum(np.isposinf(asep_array))),
        'total':            len(asep_array),
    }


def _fmt(d: dict) -> str:
    n = d['total']
    return (f"healthy={d['healthy_mono']}/{n} ({100*d['healthy_mono']/n:.1f}%), "
            f"bistable={d['bistable_finite']}/{n} ({100*d['bistable_finite']/n:.1f}%), "
            f"collapsed={d['collapsed_mono']}/{n} ({100*d['collapsed_mono']/n:.1f}%)")


def _find_latest_run_dir() -> Path:
    exp_dir = Path("outputs/fsa_v5/experiments")
    if not exp_dir.exists():
        raise SystemExit(f"No experiments dir at {exp_dir} — run from version_3/")
    runs = sorted(exp_dir.iterdir(), key=lambda p: p.name)
    runs = [p for p in runs if p.is_dir() and p.name.startswith('run')
            and (p / 'trajectory.npz').exists()]
    if not runs:
        raise SystemExit(f"No completed run dirs under {exp_dir}")
    return runs[-1]


def diagnose(run_dir: Path):
    print(f"=== {run_dir.name} ===")
    manifest_path = run_dir / 'manifest.json'
    if manifest_path.exists():
        with open(manifest_path) as f:
            m = json.load(f)
        sc = m.get('scenario', {})
        s = m.get('summary', {})
        print(f"  cost: {m.get('cost_variant')}, scenario: {sc.get('name')}, "
              f"baseline_phi: {sc.get('baseline_phi')}, T_days: {m.get('T_total_days')}")
        if 'weighted_violation_rate' in s:
            print(f"  reported weighted_violation_rate: {s['weighted_violation_rate']:.4f}")

    z = np.load(run_dir / 'trajectory.npz', allow_pickle=False)
    if 'full_phi' not in z or 'trajectory' not in z:
        raise SystemExit(f"Run dir {run_dir} lacks full_phi / trajectory keys "
                          f"in trajectory.npz; can't diagnose.")
    fphi = np.asarray(z['full_phi'], dtype=np.float64)        # (n_bins, 2)
    traj = np.asarray(z['trajectory'], dtype=np.float64)      # (n_bins, 6)
    A_traj = traj[:, 3]
    n_bins = fphi.shape[0]
    n_days = n_bins // BINS_PER_DAY
    print(f"  n_bins={n_bins}, n_days={n_days}")
    print(f"  per-bin Phi: PhiB median={np.median(fphi[:,0]):.3f} mean={fphi[:,0].mean():.3f} max={fphi[:,0].max():.3f}")
    print(f"               PhiS median={np.median(fphi[:,1]):.3f} mean={fphi[:,1].mean():.3f} max={fphi[:,1].max():.3f}")
    print(f"  A_traj: min={A_traj.min():.3f} max={A_traj.max():.3f} mean={A_traj.mean():.3f}")
    print()

    p = {k: jnp.asarray(float(v), dtype=jnp.float64)
         for k, v in TRUTH_PARAMS_V5.items()}

    def asep_at(pb_arr, ps_arr):
        return np.asarray(jax.vmap(
            lambda pb, ps: _jax_find_A_sep(pb, ps, p)
        )(jnp.asarray(pb_arr, dtype=jnp.float64),
          jnp.asarray(ps_arr, dtype=jnp.float64)))

    # 1. Per-bin (the production formulation)
    asep_bin = asep_at(fphi[:, 0], fphi[:, 1])
    cl = _classify(asep_bin)
    indicator_bin = (A_traj < asep_bin).astype(np.float64)
    print(f"[1] PER-BIN (production formulation)")
    print(f"    A_sep regime distribution: {_fmt(cl)}")
    print(f"    weighted_violation_rate (uniform per-bin weights): "
          f"{indicator_bin.mean():.4f}")
    fires = indicator_bin > 0.5
    if fires.any():
        n_collapsed_fires = int(np.sum(np.isposinf(asep_bin[fires])))
        n_bistable_fires = int(np.sum(np.isfinite(asep_bin[fires])))
        print(f"    Of {int(fires.sum())} firing bins: "
              f"{n_collapsed_fires} are A_sep=+inf (collapsed regime — vacuous), "
              f"{n_bistable_fires} are A_sep finite (real bistable violations)")
    print()

    # 2. Per-day (alternative formulation)
    days = n_bins // BINS_PER_DAY
    if days > 0:
        phi_d = fphi[:days * BINS_PER_DAY].reshape(days, BINS_PER_DAY, 2).mean(axis=1)
        A_d   = A_traj[:days * BINS_PER_DAY].reshape(days, BINS_PER_DAY).mean(axis=1)
        asep_d = asep_at(phi_d[:, 0], phi_d[:, 1])
        cl_d = _classify(asep_d)
        indicator_d = (A_d < asep_d).astype(np.float64)
        print(f"[2] PER-DAY (alternative: A_sep on daily-mean Phi)")
        print(f"    A_sep regime distribution: {_fmt(cl_d)}")
        print(f"    daily violation rate: {indicator_d.mean():.4f}  "
              f"({int(indicator_d.sum())}/{days} days)")
        print(f"    Per-day breakdown:")
        for i in range(days):
            phib, phis = phi_d[i]
            asep = asep_d[i]
            asep_str = ('-inf  HEALTHY' if np.isneginf(asep)
                        else '+inf  COLLAPSED' if np.isposinf(asep)
                        else f'{asep:6.3f} bistable')
            viol = 'VIOL' if indicator_d[i] > 0.5 else '    '
            print(f"      Day {i+1:2d}: PhiB={phib:.3f} PhiS={phis:.3f}  "
                  f"A_d={A_d[i]:.3f}  A_sep={asep_str}  {viol}")
        print()

    # 3. Active bins only (Phi above threshold)
    threshold = 0.05
    active = (fphi[:, 0] + fphi[:, 1]) > threshold
    if active.any():
        asep_act = asep_bin[active]
        A_act = A_traj[active]
        cl_a = _classify(asep_act)
        ind_a = (A_act < asep_act).astype(np.float64)
        print(f"[3] ACTIVE BINS ONLY  (sum(Phi) > {threshold}, "
              f"{int(active.sum())}/{n_bins} bins)")
        print(f"    A_sep regime distribution: {_fmt(cl_a)}")
        print(f"    active-bin violation rate: {ind_a.mean():.4f}")
        print()


def main(argv):
    if len(argv) > 1:
        run_dir = Path(argv[1])
        if not run_dir.exists():
            raise SystemExit(f"Run dir not found: {run_dir}")
    else:
        run_dir = _find_latest_run_dir()
        print(f"(No path arg; using latest: {run_dir})")
    diagnose(run_dir)


if __name__ == '__main__':
    main(sys.argv)

File: DEPRECIATED_NON_FUNCTIONAL_CODE/version_3_julia/tools/DiagnoseViolationRate.jl
# tools/DiagnoseViolationRate.jl — read-only chance-constraint diagnostic.
#
# Port of `version_3/tools/diagnose_violation_rate.py`. Three
# evaluators of the violation indicator on a completed bench run:
#
#   1. Per-bin (production formulation): `(A_traj < A_sep_per_bin)`
#      with A_sep computed at the per-bin instantaneous Phi. On
#      bursty schedules this fires vacuously during rest periods
#      (Phi=0 → A_sep=+inf → indicator=1), giving the historical
#      ~96% rate.
#   2. Per-day (alternative): A_sep at daily-mean Phi; indicator on
#      daily-mean A.
#   3. Active-bin only: indicator restricted to bins where
#      Phi_B + Phi_S > threshold.
#
# Read-only — does not run the controller. Pure CPU.

module DiagnoseViolationRate

using NPZ
using JSON3
using Printf
using ..FSAv5: FSAv5State, BimodalPhi, find_a_sep,
                BINS_PER_DAY, TRUTH_PARAMS_V5

const TRUTH = TRUTH_PARAMS_V5

function _classify(asep::AbstractVector{Float64})
    n = length(asep)
    healthy   = count(isneginf, asep)
    collapsed = count(isposinf, asep)
    bistable  = count(isfinite, asep)
    return (healthy=healthy, collapsed=collapsed, bistable=bistable, total=n)
end

function _fmt(c)
    n = c.total
    return @sprintf("healthy=%d/%d (%.1f%%), bistable=%d/%d (%.1f%%), collapsed=%d/%d (%.1f%%)",
                     c.healthy, n, 100*c.healthy/n,
                     c.bistable, n, 100*c.bistable/n,
                     c.collapsed, n, 100*c.collapsed/n)
end

"""
    diagnose(run_dir::AbstractString)

Print a 3-formulation regime breakdown for the given bench run. The
run dir must contain `trajectory.npz` with keys `full_phi` and
`trajectory`. Optionally `manifest.json` for context printing.

Mirrors `diagnose_violation_rate.diagnose` in Python.
"""
function diagnose(run_dir::AbstractString)
    println("=== ", basename(run_dir), " ===")

    manifest_path = joinpath(run_dir, "manifest.json")
    if isfile(manifest_path)
        m = JSON3.read(read(manifest_path, String))
        sc = get(m, :scenario, Dict())
        s  = get(m, :summary, Dict())
        println("  cost: $(get(m, :cost_variant, "?"))  scenario: $(get(sc, :name, "?"))  T_days: $(get(m, :T_total_days, "?"))")
        if haskey(s, :weighted_violation_rate)
            @printf "  reported weighted_violation_rate: %.4f\n" s.weighted_violation_rate
        end
    end

    z      = NPZ.npzread(joinpath(run_dir, "trajectory.npz"))
    fphi   = Float64.(z["full_phi"])         # (n_bins, 2)
    traj   = Float64.(z["trajectory"])       # (n_bins, 6)
    A_traj = traj[:, 4]
    n_bins = size(fphi, 1)
    n_days = n_bins ÷ BINS_PER_DAY

    @printf "  n_bins=%d, n_days=%d\n" n_bins n_days
    @printf "  per-bin Phi: PhiB median=%.3f mean=%.3f max=%.3f\n" median(fphi[:,1]) mean(fphi[:,1]) maximum(fphi[:,1])
    @printf "               PhiS median=%.3f mean=%.3f max=%.3f\n" median(fphi[:,2]) mean(fphi[:,2]) maximum(fphi[:,2])
    @printf "  A_traj: min=%.3f max=%.3f mean=%.3f\n" minimum(A_traj) maximum(A_traj) mean(A_traj)
    println()

    p = TRUTH

    # [1] Per-bin (production)
    asep_bin = [find_a_sep(BimodalPhi(fphi[t, 1], fphi[t, 2]), p)
                 for t in 1:n_bins]
    cl = _classify(asep_bin)
    ind_bin = Float64.(A_traj .< asep_bin)
    println("[1] PER-BIN (production formulation)")
    println("    A_sep regime distribution: ", _fmt(cl))
    @printf "    weighted_violation_rate: %.4f\n" mean(ind_bin)
    fires = ind_bin .> 0.5
    n_fires = sum(fires)
    if n_fires > 0
        n_collapsed = count(isposinf, asep_bin[fires])
        n_finite    = count(isfinite, asep_bin[fires])
        @printf "    Of %d firing bins: %d are A_sep=+inf (vacuous), %d finite (real bistable)\n" n_fires n_collapsed n_finite
    end
    println()

    # [2] Per-day (alternative)
    if n_days > 0
        phi_d = reshape(fphi[1:n_days*BINS_PER_DAY, :], BINS_PER_DAY, n_days, 2)
        phi_d_mean = vec(mean(phi_d, dims=1))   # length n_days * 2 then reshape
        phi_d_mat  = reshape(phi_d_mean, n_days, 2)
        A_d_mat    = vec(mean(reshape(A_traj[1:n_days*BINS_PER_DAY], BINS_PER_DAY, n_days), dims=1))
        asep_d     = [find_a_sep(BimodalPhi(phi_d_mat[i, 1], phi_d_mat[i, 2]), p)
                       for i in 1:n_days]
        cl_d = _classify(asep_d)
        ind_d = Float64.(A_d_mat .< asep_d)
        println("[2] PER-DAY (alternative: A_sep on daily-mean Phi)")
        println("    A_sep regime distribution: ", _fmt(cl_d))
        @printf "    daily violation rate: %.4f  (%d/%d days)\n" mean(ind_d) Int(sum(ind_d)) n_days
        println("    Per-day breakdown:")
        for i in 1:n_days
            ai   = asep_d[i]
            tag  = isneginf(ai) ? "-inf  HEALTHY" :
                    isposinf(ai) ? "+inf  COLLAPSED" :
                    @sprintf "%6.3f bistable" ai
            viol = ind_d[i] > 0.5 ? "VIOL" : "    "
            @printf "      Day %2d: PhiB=%.3f PhiS=%.3f A_d=%.3f A_sep=%s  %s\n"  i phi_d_mat[i,1] phi_d_mat[i,2] A_d_mat[i] tag viol
        end
        println()
    end

    # [3] Active bins only
    threshold = 0.05
    active = (fphi[:, 1] .+ fphi[:, 2]) .> threshold
    if any(active)
        asep_act = asep_bin[active]
        A_act    = A_traj[active]
        cl_a     = _classify(asep_act)
        ind_a    = Float64.(A_act .< asep_act)
        @printf "[3] ACTIVE BINS ONLY  (sum(Phi) > %.2f, %d/%d bins)\n" threshold sum(active) n_bins
        println("    A_sep regime distribution: ", _fmt(cl_a))
        @printf "    active-bin violation rate: %.4f\n" mean(ind_a)
        println()
    end
end

# ── Main entry ─────────────────────────────────────────────────────────────

function main(argv::Vector{<:AbstractString} = ARGS)
    run_dir = isempty(argv) ?
        # Find the latest run dir under outputs/fsa_v5/experiments/
        let exp = joinpath(pwd(), "outputs", "fsa_v5", "experiments")
            isdir(exp) || error("No experiments dir at $exp")
            runs = sort([joinpath(exp, e) for e in readdir(exp) if startswith(e, "run") &&
                          isfile(joinpath(exp, e, "trajectory.npz"))])
            isempty(runs) ? error("no run dirs with trajectory.npz") : last(runs)
        end :
        argv[1]
    diagnose(run_dir)
end

# Helpers
using Statistics: mean, median

export diagnose, main

end # module DiagnoseViolationRate
